Interop/ONNX/src/main/java/org/tribuo/interop/onnx/extractors/BERTFeatureExtractor.java

```diff
@@ -0,0 +1,988 @@
+        int size = tokens.size() + 2; // for [CLS] and [SEP]
+        long[] curTokenIds = new long[size];
+
+        curTokenIds[0] = tokenIDs.get(classificationToken);
+        int i = 1;
+        for (String token : tokens) {
+            Integer id = tokenIDs.get(token);
+            if (id == null) {
+                curTokenIds[i] = tokenIDs.get(unknownToken);
+            } else {
+                curTokenIds[i] = id;
+            }
+            i++;
+        }
+        curTokenIds[i] = tokenIDs.get(separatorToken);
+
+        return OnnxTensor.createTensor(env,new long[][]{curTokenIds});
+    }
+
+    /**
+     * Creates a constant tensor filled with the specified value.
+     * @param size The size of tensor to create.
+     * @return The tensor.
+     * @throws OrtException if it failed to create the tensor.
+     */
+    private OnnxTensor createTensor(int size, long value) throws OrtException {
+        long[] array = new long[size];
+        Arrays.fill(array,value);
+        return OnnxTensor.createTensor(env,new long[][]{array});
+    }
+
+    /**
+     * Reads bertDim values off the buffer, throwing {@link java.nio.BufferUnderflowException} if we exceed the buffer.
+     * <p>
+     * Advances the state of the buffer.
+     * @param buffer The float buffer to read.
+     * @param bertDim The number of values to read.
+     * @return The features.
+     */
+    private static double[] extractFeatures(FloatBuffer buffer, int bertDim) {
+        double[] features = new double[bertDim];
+        float[] floatArr = new float[bertDim];
+        buffer.get(floatArr);
+        for (int i = 0; i < floatArr.length; i++) {
+            features[i] = floatArr[i];
+        }
+        return features;
+    }
+
+    /**
+     * Reads bertDim values off the buffer, throwing {@link java.nio.BufferUnderflowException} if we exceed the buffer.
+     * <p>
+     * Advances the state of the buffer.
+     * <p>
+     * Adds the feature values to the values array.
+     * @param buffer The float buffer to read.
+     * @param bertDim The number of values to read.
+     * @param values The values to add.
+     */
+    private static void addFeatures(FloatBuffer buffer, int bertDim, double[] values) {
+        float[] floatArr = new float[bertDim];
+        buffer.get(floatArr);
+        for (int i = 0; i < floatArr.length; i++) {
+            values[i] += floatArr[i];
+        }
+    }
+
+    /**
+     * Passes the tokens through BERT, replacing any unknown tokens with the [UNK] token.
+     * <p>
+     * The features of the returned example are dense, and come from the [CLS] token.
+     * <p>
+     * Throws {@link IllegalStateException} if the BERT model failed to produce an output.
+     * @param tokens The input tokens. Should be tokenized using the Tokenizer this BERT expects.
+     * @return A dense example representing the pooled output from BERT for the input tokens.
+     */
+    public Example<T> extractExample(List<String> tokens) {
```

Comment:
Is there a check somewhere along this path to make sure `tokens` doesn't exceed max length?

Reply:
Nope, not on this path. I wasn't sure if truncating it or throwing an `IllegalArgumentException` was better. This is supposed to be pretokenized input.

Reply:
I added `IllegalArgumentException` to the methods that accept a list as the programmer could check if it's longer than `getMaxLength` and deal with it themselves.